File: 743-network-delay-time/743-network-delay-time.java
class Solution {
    
    public int networkDelayTime(int[][] times, int n, int k) {
        if(n < 2) return 0;
        List<List<int[]>> graph = new ArrayList<>();
        for(int i = 0; i <= n; i++) graph.add(new ArrayList<int[]>());
        
        for(int[] edge : times){
            graph.get(edge[0]).add(new int[]{edge[1],edge[2]});
        }
        int[] distance = new int[n+1];
        Arrays.fill(distance,Integer.MAX_VALUE);
        distance[k] = 0;
        boolean[] visited = new boolean[n+1];
        Queue<int[]> queue = new PriorityQueue<>((a,b)->a[1] - b[1]);
        queue.add(new int[]{k,0});
        while(!queue.isEmpty()){
            int[] curr = queue.poll();
            visited[curr[0]] = true;
            for(int[] adj : graph.get(curr[0])){
                if(!visited[adj[0]]){
                    int currDist = curr[1] + adj[1];
                    if(distance[adj[0]] > currDist){
                        distance[adj[0]] = currDist;
                        queue.add(new int[]{adj[0],currDist});
                    }
                }
            }
        }
        int time = 0;
        for(int i = 1; i  <= n ; i++){
            int t = distance[i];
            if(t == Integer.MAX_VALUE) return -1;
            time = Math.max(time,t);
        }
        return time;
    }
}

File: 743-network-delay-time/743-network-delay-time.py
class Solution:
    def networkDelayTime(self, times: List[List[int]], n: int, k: int) -> int:
        
        graph = defaultdict(list)
        for edge in times:
            graph[edge[0]].append((edge[1],edge[2]))
        queue = deque()
        queue.append((k,0))
        MAX = 10**20
        dist = [MAX] * (n + 1)
        dist[0] = -1
        dist[k] = 0
        while(len(queue) > 0):
            curr = queue.popleft()
            for adj in graph[curr[0]]:
                if(adj[1] + curr[1] < dist[adj[0]]):
                    queue.append((adj[0],adj[1] + curr[1]))
                    dist[adj[0]] = adj[1] + curr[1]        
        
        ans = max(dist)
        
        return -1 if ans == MAX else ans